src/pdu.cpp

```diff
@@ -85,6 +85,20 @@ uint32_t PDU::size() const {
     return sz;
 }
 
+uint32_t PDU::advertised_size() const {
+    uint32_t sz = header_size() + trailer_size();
+    const PDU* ptr(inner_pdu_);
+    while (ptr) {
+        sz += ptr->advertised_size();
+        if(ptr->pdu_type() != IP) {
```

Comment:
This check shouldn't be here. This has to work no matter what PDUs a packet contains. I think a better and generic solution to this is to have the default implementation return `header_size() + trailer_size() + inner_pdu()->size()`  (this last one obviously only if there's an inner PDU). This way you're defining the advertised PDU size to be the size of the current PDU + whatever the next PDU tells you. So if the next PDU is IP, then it will simply return its total length without going further down the chain.

Reply:
Sorry, I mean `inner_pdu()->advertised_size()` here.

Reply:
Well the IP::tot_len() contains the lenght of all inner PDUs does it not?

That was my understanding after looking at the IP header in wireshark.

Reply:
Exactly, which is why IP can get away with just returning `tot_len` as its advertised size vs the rest of the PDUs which need to keep going down the PDU chain to find out.

Reply:
I see the check is no longer needed, since we IP::advertised_size() will be the recursion anchor.

I'll fix it.

Reply:
Made the changes, just waiting for your continuous integration.